src/app/modules/products/pages/product/product.page.ts

```diff
@@ -12,27 +19,90 @@ import { Product } from "@shared/models/classes/product.class";
+
+    get products$(): Observable<Product[]> {
+        return this.products.asObservable();
+    }
 
     constructor(
         private productService: ProductApiService,
         private route: ActivatedRoute,
+        private fb: FormBuilder,
     ) {}
 
     ngOnInit(): void {
         this.featuredProducts$ = this.productService.getFeaturedProducts();
 
-        this.subscriptions.add(this.route.params
-            .subscribe(
-                ({productType}) => {
-                    this.products$ = this.productService.getProductsByType(productType);
-                }
+        this.subscriptions.add(
+            this.route.params.subscribe(
+                ({productType}) =>
+                    this.subscriptions.add(
+                        this.productService.getProductsByType(productType).subscribe(
+                            products => {
+                                this.products.next(products);
+                                this.updateFilterForm(products, productType);
+                            })
+                    )
+            )
+        );
+
+        this.subscriptions.add(
+            this.filterForm.valueChanges.subscribe(
+                changes => this.updateActiveFilters(changes)
             )
         );
     }
 
     ngOnDestroy(): void {
         this.subscriptions.unsubscribe();
     }
+
+    private updateFilterForm(products: Product[], productType: ProductType): void {
+        this.filterableProperties
+            ?.forEach(p => this.filterForm.removeControl(p.key));
+
+        this.filterableProperties = (this.getFilterableProperties(products, productType));
+
+        this.filterableProperties
+            .forEach(p => this.filterForm.addControl(p.key, new FormControl()));
+    }
+
+    private getFilterableProperties(products: Product[], productType: ProductType): FilterType[] {
+        const filters = products
+            .map(p => p.filterableAttributes)
+            .reduce((arr, fa) => arr.concat(fa))
+            .map(fa => {
+                return {
+                    [fa.attribute]: {
+                        displayName: fa.displayName,
+                        key: fa.attribute,
+                        options: getUniqueValuesOfKey(products, fa.attribute),
+                    }
+                }
+            })
+            .reduce((union, obj) => {
+                return { ...union, ...obj }
+            });
+
+        if (productType != "all") delete filters["productType"];
+
+        return Object.values(filters);
+    }
+
+    private updateActiveFilters(changes: any): void {
+        this.activeFilters = Object.keys(changes)
+            .filter(key => changes[key] != null)
+            .map(key => {
+                const active = [...changes[key]];
+                return active.length == 0 ? {} : { [key]: active };
+            })
+            .reduce((union, obj) => {
+                return { ...union, ...obj }
+            }, {});
```

Comment:
you can remove the filter part and add a check for !=null in the map portion of this block.

This is also a good place to use a getter/setter combination

Reply:
I gave this a shot, but TS requires that the return value from a getter is exactly the same as the input value for a setter, so I couldn't see how to break this in to two parts. You'd either need to require all the callers of setters to process the input before they passed it in (which would be, essentially, this method here), or do everything in the setter, but lose type checking by having to return `any` (the type of `changes` from the form), and have to do a lot more error checking.

If I've misunderstood, let me know,